File: application-service/src/main/java/com/hireconnect/application/service/ApplicationService.java
package com.hireconnect.application.service;

import java.time.LocalDate;
import java.util.List;

import com.hireconnect.application.dto.ApplicationRequest;
import com.hireconnect.application.dto.ApplicationResponse;

public interface ApplicationService {

    ApplicationResponse submitApplication(ApplicationRequest request);

    List<ApplicationResponse> getByCandidate(Integer candidateId);

    List<ApplicationResponse> getByJob(Integer jobId);

    List<ApplicationResponse> getByStatus(String status);

    List<ApplicationResponse> getByAppliedDateRange(LocalDate startDate, LocalDate endDate);

    ApplicationResponse getById(Integer applicationId);

    ApplicationResponse updateStatus(Integer applicationId, String status);

    ApplicationResponse withdrawApplication(Integer applicationId);

    int countByJob(Integer jobId);
}

File: frontend/src/app/core/services/application.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { API_ENDPOINTS } from '../constants/api.constants';
import { ApplicationRequest, ApplicationResponse, StatusUpdateRequest } from '../models/application.models';

@Injectable({ providedIn: 'root' })
export class ApplicationService {
  private readonly http = inject(HttpClient);

  submit(payload: ApplicationRequest): Observable<ApplicationResponse> {
    return this.http.post<ApplicationResponse>(API_ENDPOINTS.applications, payload);
  }

  getById(applicationId: number): Observable<ApplicationResponse> {
    return this.http.get<ApplicationResponse>(`${API_ENDPOINTS.applications}/${applicationId}`);
  }

  getByCandidate(candidateId: number): Observable<ApplicationResponse[]> {
    return this.http.get<ApplicationResponse[]>(`${API_ENDPOINTS.applications}/candidate/${candidateId}`);
  }

  getByJob(jobId: number): Observable<ApplicationResponse[]> {
    return this.http.get<ApplicationResponse[]>(`${API_ENDPOINTS.applications}/job/${jobId}`);
  }

  countByJob(jobId: number): Observable<number> {
    return this.http.get<number>(`${API_ENDPOINTS.applications}/job/${jobId}/count`);
  }

  searchByStatus(status: string): Observable<ApplicationResponse[]> {
    return this.http.get<ApplicationResponse[]>(API_ENDPOINTS.applications, {
      params: new HttpParams().set('status', status),
    });
  }

  searchByDateRange(appliedFrom: string, appliedTo: string): Observable<ApplicationResponse[]> {
    return this.http.get<ApplicationResponse[]>(API_ENDPOINTS.applications, {
      params: new HttpParams().set('appliedFrom', appliedFrom).set('appliedTo', appliedTo),
    });
  }

  updateStatus(applicationId: number, payload: StatusUpdateRequest): Observable<ApplicationResponse> {
    return this.http.patch<ApplicationResponse>(`${API_ENDPOINTS.applications}/${applicationId}/status`, payload);
  }

  withdraw(applicationId: number): Observable<ApplicationResponse> {
    return this.http.patch<ApplicationResponse>(`${API_ENDPOINTS.applications}/${applicationId}/withdraw`, {});
  }
}
